library/src-bootstrapped/scala/quoted/util/ExprMap.scala

```diff
@@ -0,0 +1,158 @@
+package scala.quoted.util
+
+import scala.quoted._
+
+trait ExprMap {
+
+  /** Map an expression `e` with a type `tpe` */
+  def map[T](e: Expr[T])(given qctx: QuoteContext, tpe: Type[T]): Expr[T]
+
+  /** Map subexpressions an expression `e` with a type `tpe` */
+  def mapChildren[T](e: Expr[T])(given qctx: QuoteContext, tpe: Type[T]): Expr[T] = {
+    import qctx.tasty.{_, given}
+    class MapChildren() {
```

Comment:
How different is the class from `TreeMap`? Can it be implemented on top of that?

Reply:
The extra type argument makes the methods signatures different (not all) the logic in each one has to handle the extra type parameter.

I could not manage to implement it on to of the other one.